### DISCUSSION TITLE
return value modes

### DISCUSSION THREAD

**Discussion by tjhance** [verus-lang/verus#109] [Language design]
### Context

By default, the return value of a function has the same mode as the function. This can be overridden with an attribute:

```rust
#[exec]
#[verifier(returns(proof))]
fn foo() -> T
```

(Apparently, there was a technical challenge with making it an attribute on the return type (rust type's don't have attributes?) so it's on the function instead.)

To call `foo`, you'd write something like:

```rust
#[proof] let x = foo();
```

Here, the `#[proof]` attribute indicates that the _variable being declared_ (that is, `x`) should have mode `proof`. Note that the call to `foo()` is still `exec`-mode.

In order to handle returning tuples where some fields have different modes, Verus also introduced two special types `Proof<...>` and `Spec<...>`.

You'd do:

```rust
#[exec]
fn foo() -> (A, Proof<B>)
```

Note that the `returns(proof)` attribute is now gone.

The first argument of the returned tuple would have mode `exec` (the default mode, that is, same mode as the function) and the second argument has mode `proof`. At the call site, it looks like this:

```rust
let (a, Proof(b)) = foo();
```

### Some opinions

The new style, with the `Spec` and `Proof` types, look a lot cleaner than the attribute-style, even for functions that only return one value. Compare:

```rust
#[exec]
#[verifier(returns(proof))]   // attribute is visually decoupled from the return type
fn foo() -> T

#[exec]
fn foo() -> Proof<T>          // now the 'Proof' annotation is right there with the type
```

Also compare the call sites:

```rust
#[proof] let x = foo();       // looks like the entire line is `proof` but the `foo` call is actually `exec`

let Proof(x) = foo();         // it's clear that the 'Proof' annotation is associated with the new variable
```

Note: this currently doesn't work as well with `Spec`, but this is a technical limitation that should be fixable with some `rustc` changes.

### My proposal

 * Remove the `#[verifier(returns(...))]` attribute.
 * Use `Proof` and `Spec` exclusively to determine the modes of returned variables.
 * If a statement has an attribute like `#[proof]` or `#[spec]`, then that should restrict the mode of the entire statement.

**Comment by Chris-Hawblitzel** [comment]
This shows an example of an `exec`-mode function `foo`, returning a `proof` value, called from an `exec` context.  Could you write out the other scenarios as well?  I'd expect that this proposal doesn't affect calls to `spec`-mode functions, but there are various combinations for calls to `proof`-mode functions that would be interesting, depending on whether the function returns `proof` or `spec` values and whether it's called from an `exec` or `proof` context.

**Reply by tjhance** [reply]
```rust
// exec fn returning proof:
let Proof(t) = exec_call();

// exec fn returning spec:
let Spec(t) = exec_call();    // assuming we get this working

// proof fn returning spec
let Spec(t) = proof_call();
```

I think proof-returning-spec is worth discussing ... I hadn't really thought about it until now, but I assume that proof-returning-spec is the common case (rather than proof-returning-proof), which means it's a little awkward if you have to write `Spec(t)` all the time. Then again, lemmas returning values isn't _that_ common, either, and it's not like it's _more_ awkward than before. But I still think it's worth examining closely.
